src/search/search.c

```diff
@@ -0,0 +1,87 @@
+/* Copyright (c) 2013 Jonathan Klee
+
+This file is part of ngp.
+
+ngp is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+ngp is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with ngp.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <string.h>
+
+#include "../entry.h"
+#include "../list.h"
+#include "../options.h"
+#include "search.h"
+
+
+void do_ngp_search(struct search_t *search);
+void do_ag_search(struct search_t *search);
+void do_git_search(struct search_t *search);
+
+struct search_t * create_search( struct options_t *options )
+{
+    struct result_t* result = calloc(1, sizeof(*result));
+    result->entries = NULL;
+    result->start = result->entries;
+    result->nbentry = 0;
+
+    struct search_t *search = calloc(1, sizeof(*search));
+    search->result = result;
+    search->status = 1;
+
+    search->options = options;
+
+    return search;
+}
+
+void do_search(struct search_t *search)
+{
+    switch (search->options->search_type) {
+        case NGP_SEARCH:
+            return do_ngp_search(search);
+
+        case AG_SEARCH:
+            return do_ag_search(search);
+
+        case GIT_SEARCH:
+            return do_git_search(search);
+
+        default:
+            break;
+    }
+
```

Comment:
Instead of that switch/case, I would rather set up an inheritance relationship between searches.
I would do a parent search_t object and native_search_t, ag_search_t, git_search_t child objects in the same way entry_t is currently parenting line_t and file_t.
What do you think ?

Reply:
I thought of that. And if it were a C++ code I definitely would have implemented it this way. But since it is plain C, and in the end it's only 1 function I wanted to avoid the IMHO more complicated code.

I will change it you insist, but I think C++ is more suitable for creating an inheritance relationship.

Reply:
yes sure but I would rather keep the whole code consistent

Reply:
@jonathanklee 
I just wanted to implement the search vtable to make it more consistent with file and line. But, in order to keep the interface for `search_t` simple, I would just move the switch statement  from `do_search` to `create_search` and set a callback in the specific versions of `create_search`.
IMHO there is nothing gained by that.
And changing the interface of  `search_t` to something like `search->do_search(search)` would in my opinion also not be beneficial for consistency. Because the constructor call (`create_search`) will still be called explicitly. 
 
What do you think? Do you have a better idea?